Foundation/NSTask.swift

```diff
@@ -7,13 +7,127 @@
 // See http://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
 //
 
+import CoreFoundation
+
+#if os(OSX) || os(iOS)
+    import Darwin
+#elseif os(Linux)
+    import Glibc
+#endif
 
 public enum NSTaskTerminationReason : Int {
-    
     case Exit
     case UncaughtSignal
 }
 
+private func WEXITSTATUS(status: CInt) -> CInt {
+    return (status >> 8) & 0xff
+}
+
+
+private var processMap : [pid_t : NSTask] = [:]
+private var processLaunchLock = NSLock()
+
+private var managerThreadSetupOnceToken = pthread_once_t()
+private var threadID = pthread_t()
+
+private let semaphore = sem_open("_NSTaskRunLoopSemaphore", O_CREAT, 0o777, 0)
+
+@noreturn private func managerThread(x: UnsafeMutablePointer<Void>) -> UnsafeMutablePointer<Void> {
+    
+    while true {
+        sem_wait( semaphore )
+        
+        // Block for any child process to finish
+        
+        var status = Int32()
+        let pid = waitpid( 0, &status, 0 )
+        
+        // Lock
+        
+        processLaunchLock.lock()
+        
+        // Get the associated NSTask object
+        
+        if let task = processMap[pid] {
+            
+            // Get the exit status (TODO: handle the different kinds of exits)
+            
+            task.terminationStatus = WEXITSTATUS( status )
+            
+            // If a termination handler has been set, invoke it on a background thread
+            
+            if task.terminationHandler != nil {
+                var threadID = pthread_t()
+                pthread_create(&threadID, nil, { (context) -> UnsafeMutablePointer<Void> in
+                    
+                    let unmanagedTask : Unmanaged<NSTask> = Unmanaged.fromOpaque(context)
+                    let task = unmanagedTask.takeRetainedValue()
+                    
+                    task.terminationHandler!( task )
+                    return context
+                    
+                    }, UnsafeMutablePointer<Void>(Unmanaged.passRetained(task).toOpaque()))
+            }
+            
+            // Set the running flag to false
+            
+            task.running = false
+            
+            // Invalidate the source and wake up the run loop, if it's available
+            
+            CFRunLoopSourceInvalidate(task.runLoopSource)
+            if let runLoop = task.runLoop {
+                CFRunLoopWakeUp(runLoop._cfRunLoop)
+            }
+        }
+        
+        // Clear out the NSTask from the process table
+        
+        processMap[pid] = nil
+        
+        processLaunchLock.unlock()
+    }
+}
+
+private func runLoopCallback(context : UnsafeMutablePointer<Void>) -> Void {}
+
+private func managerThreadSetup() -> Void {
+    pthread_create(&threadID, nil, managerThread, nil)
+}
+
+
+// Equal method for run loop source
+private func nstaskIsEqual(a : UnsafePointer<Void>, b : UnsafePointer<Void>) -> DarwinBoolean {
```

Comment:
DarwinBoolean does not compile on linux so we should probably fix CFRunLoopSourceContext